instrumentation/mongo/src/main/java/brave/mongo/TraceMongoCommandListener.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2013-2020 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.mongo;
+
+import brave.Span;
+import brave.Tracer;
+import com.mongodb.event.CommandFailedEvent;
+import com.mongodb.event.CommandListener;
+import com.mongodb.event.CommandStartedEvent;
+import com.mongodb.event.CommandSucceededEvent;
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+import org.bson.json.JsonWriterSettings;
+
+import java.net.InetSocketAddress;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A Mongo command listener that will report via Brave how long each command takes.
+ *
+ * <p>To use it, call <code>.addCommandListener(TraceMongoCommandListener.builder().tracer(tracer).build())</code>
+ * on the {@link com.mongodb.MongoClientOptions} or {@link com.mongodb.MongoClientSettings} object that is used to
+ * create the {@code MongoClient} to be instrumented.
+ */
+public class TraceMongoCommandListener implements CommandListener {
+  private final Tracer tracer;
+  private final int maxAbbreviatedCommandLength;
+  private final JsonWriterSettings jsonWriterSettings;
+  private final Map<Integer, Span> activeSpans = new ConcurrentHashMap<>();
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  protected TraceMongoCommandListener(final Builder builder) {
+    tracer = Objects.requireNonNull(builder.tracer, "tracer cannot be null");
+    maxAbbreviatedCommandLength = builder.maxAbbreviatedCommandLength;
+    jsonWriterSettings = JsonWriterSettings.builder()
+      .maxLength(Math.max(0, builder.maxAbbreviatedCommandLength))
+      .build();
+  }
+
+  @Override public void commandStarted(final CommandStartedEvent event) {
+    final Span span = tracer.nextSpan();
+    if (span.isNoop()) {
+      return;
+    }
+    activeSpans.put(event.getRequestId(), span);
+
+    final String commandName = event.getCommandName();
+    final String databaseName = event.getDatabaseName();
+    final BsonDocument command = event.getCommand();
+    final Optional<String> collectionName = getCollectionName(command);
+
+    span.name(getSpanName(commandName, collectionName))
+      .kind(Span.Kind.CLIENT)
+      .remoteServiceName(databaseName)
+      .tag("db.type", "mongo")
```

Comment:
we don't use any tag like db.type elsewhere (ex mysql) so I don't think we'd do that here.

also we need to check the tags are like existing mongo instrumentation people in zipkin use. please ask the people on the referring issue as this type of default policy is something to consider carefully. particularly the tag names ex mongo vs mongodb.

@kojilin @huydx do you use mongo at LINE?

Reply:
ps it may not be obvious why we never add "db.type" or "rpc.type" etc. main thing is every tag adds expense, indexing load and size.

Also, since there is no abstraction here, ex "db.name", you can easily find tag associated by searching for the concrete tag also added. ex in zipkin presence of tag is a valid search ex just "mongodb.database". if this convention was taken from another library, someone can also change the policy to backfill "db.name" for looking for the other tag. Later, we can add a parser here which would allow policy overrides to be easier also, but for now, I think least count of most usable tags is how to start.

Reply:
ps while it isn't as usual for tags to match on name between metrics and tracing, names can be correlated. Brave is frequently used with micrometer, so you can look at what they tag https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsCommandListener.java#L62

Reply:
Yes, we are using MongoDB. IMHO, databaseName usually provide enough information. or like MySQL's code, `mongodb-$databasename`?

I have a question when we try to build a similar thing, we found MongoDB  sometimes doesn't use the same thread as client caller. It means the span may be not belong to parent correctly.  Or recent MongoDB driver doesn't have this problem?

Reply:
@adriancole do we have other instrumentation where the chance of losing the parent span is built in? This seems like we need a huge disclaimer.

Reply:
specifically on this parent-attribution thing, we either wrap to stash the invocation context (ex `CurrentTraceContext.wrap()`) or we use a property bag given by the framework https://github.com/openzipkin/brave/blob/master/instrumentation/okhttp3/src/main/java/brave/okhttp3/TracingCallFactory.java#L62

Reply:
>  I have a question when we try to build a similar thing, sometimes we found MongoDB is not using the same thread as client caller. It means the span may not belong to parent correctly. Or recent MongoDB driver doesn't have this problem? But maybe ok because of it only happens occasionally.

@kojilin
For the synchronous/legacy client, I believe the answer is that only some maintenance tasks are done on a thread other than the caller's thread:
- [Cursor cleaning](https://github.com/mongodb/mongo-java-driver/blob/67c9f738ae44bc15befb822644e7266634c7dcf5/driver-legacy/src/main/com/mongodb/MongoClient.java#L802)
- [Connection pool maintenance](https://github.com/mongodb/mongo-java-driver/blob/67c9f738ae44bc15befb822644e7266634c7dcf5/driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java#L95)

So I think it's okay that spans from those threads are not attached to any parent span.

---

The reactive/async client on the other hand uses threads for its async completion handlers (so `onCommandStarted` and `onCommandSucceeded`/`Failed` may get called from background threads and also not necessarily from the same thread) but I believe that it is possible to set a custom `StreamFactoryFactory` on the `MongoClientSettings.Builder` which can in theory propagate the tracing context correctly between those handlers.

(It's likely that I will punt on supporting the reactive/async client in this first version to try to keep the scope/complexity at bay.)

Reply:
great research. in the code comments, please highlight the cursor and connection pool links and/or have unit tests if that's something easy to do.